lib/std/crypto/Certificate.zig

```diff
@@ -0,0 +1,724 @@
+        .message_slice = .{ .start = certificate.slice.start, .end = tbs_certificate.slice.end },
+        .pub_key_algo = pub_key_algo,
+        .pub_key_slice = pub_key,
+        .validity = .{
+            .not_before = not_before_utc,
+            .not_after = not_after_utc,
+        },
+    };
+}
+
+pub fn verify(subject: Certificate, issuer: Certificate) !void {
+    const parsed_subject = try subject.parse();
+    const parsed_issuer = try issuer.parse();
+    return parsed_subject.verify(parsed_issuer);
+}
+
+pub fn contents(cert: Certificate, elem: der.Element) []const u8 {
+    return cert.buffer[elem.slice.start..elem.slice.end];
+}
+
+pub fn parseBitString(cert: Certificate, elem: der.Element) !der.Element.Slice {
+    if (elem.identifier.tag != .bitstring) return error.CertificateFieldHasWrongDataType;
+    if (cert.buffer[elem.slice.start] != 0) return error.CertificateHasInvalidBitString;
+    return .{ .start = elem.slice.start + 1, .end = elem.slice.end };
+}
+
+/// Returns number of seconds since epoch.
+pub fn parseTime(cert: Certificate, elem: der.Element) !u64 {
+    const bytes = cert.contents(elem);
+    switch (elem.identifier.tag) {
+        .utc_time => {
+            // Example: "YYMMDD000000Z"
+            if (bytes.len != 13)
+                return error.CertificateTimeInvalid;
+            if (bytes[12] != 'Z')
+                return error.CertificateTimeInvalid;
+
+            return Date.toSeconds(.{
+                .year = @as(u16, 2000) + try parseTimeDigits(bytes[0..2].*, 0, 99),
+                .month = try parseTimeDigits(bytes[2..4].*, 1, 12),
+                .day = try parseTimeDigits(bytes[4..6].*, 1, 31),
+                .hour = try parseTimeDigits(bytes[6..8].*, 0, 23),
+                .minute = try parseTimeDigits(bytes[8..10].*, 0, 59),
+                .second = try parseTimeDigits(bytes[10..12].*, 0, 59),
+            });
+        },
+        .generalized_time => {
+            // Examples:
+            // "19920521000000Z"
+            // "19920622123421Z"
+            // "19920722132100.3Z"
+            if (bytes.len < 15)
+                return error.CertificateTimeInvalid;
+            return Date.toSeconds(.{
+                .year = try parseYear4(bytes[0..4]),
+                .month = try parseTimeDigits(bytes[4..6].*, 1, 12),
+                .day = try parseTimeDigits(bytes[6..8].*, 1, 31),
+                .hour = try parseTimeDigits(bytes[8..10].*, 0, 23),
+                .minute = try parseTimeDigits(bytes[10..12].*, 0, 59),
+                .second = try parseTimeDigits(bytes[12..14].*, 0, 59),
+            });
+        },
+        else => return error.CertificateFieldHasWrongDataType,
+    }
+}
+
+const Date = struct {
+    /// example: 1999
+    year: u16,
+    /// range: 1 to 12
+    month: u8,
+    /// range: 1 to 31
+    day: u8,
+    /// range: 0 to 59
+    hour: u8,
+    /// range: 0 to 59
+    minute: u8,
+    /// range: 0 to 59
+    second: u8,
```

Comment:
question

Is there a reason to always use `u8`?
e.g.
```zig
    /// range: 1 to 12
    month: u4,
    /// range: 1 to 31
    day: u5,
    /// range: 0 to 59 <<< should it be 23?
    hour: u5,
    /// range: 0 to 59
    minute: u6,
    /// range: 0 to 60
    second: u6,
```

Reply:
I'm looking forward to specific integer ranges #3806